src/bt_peer_connection.cpp

```diff
@@ -358,6 +358,51 @@ namespace libtorrent {
 		stats_counters().inc_stats_counter(counters::num_outgoing_dht_port);
 	}
 
+	void bt_peer_connection::send_piece_message(message_type const type,
+			counters::stats_counter_t const counter,
+			piece_index_t const index)
+	{
+		char msg[] = { 0,0,0,5, static_cast<char>(type), 0, 0, 0, 0 };
+		char* ptr = msg + 5;
+		detail::write_int32(static_cast<int>(index), ptr);
+		send_buffer(msg, sizeof(msg));
+
+		stats_counters().inc_stats_counter(counter);
+	}
+
+	void bt_peer_connection::send_simple_message(message_type const type,
+			counters::stats_counter_t const counter)
+	{
+		TORRENT_ASSERT(m_sent_handshake);
+		TORRENT_ASSERT(m_sent_bitfield);
+
+		char msg[] = { 0,0,0,1,static_cast<char>(type) };
+		send_buffer(msg, sizeof(msg));
+
+		stats_counters().inc_stats_counter(counter);
+	}
+
+	void bt_peer_connection::send_request_message(message_type const type,
```

Comment:
it seems like you could merged all these 3 functions into a single template. They all has 0, 1 or 3 int32 arguments.